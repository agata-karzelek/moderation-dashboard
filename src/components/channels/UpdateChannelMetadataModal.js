import React, { useState, useEffect } from "react";
import { DialogTitle, Dialog, Typography, IconButton } from "@material-ui/core";
import CloseIcon from "@material-ui/icons/Close";
import { useStyles } from "../../style/createModerationModal";
import { editChannelMetadata } from "../../services/pubnub";
import SnackBar from "../core/SnackBar";
import { showError } from "../../utils/helpers";
import { useFormik } from "formik";
import AddChannelForm from "./AddChannelForm";
import { channelValidationSchema } from "./AddChannelMetadataModal";

export default function UpdateChannelMetadataModal(props) {
  const { pubnubObject, open, setOpen, data, channelUpdated } = props;
  const [name, setName] = useState("");
  const [description, setDescription] = useState(null);
  const [channelID, setChannelID] = useState("");
  const [updateChannelAlert, setUpdateChannelAlert] = useState({
    success: { status: false, msg: "" },
    error: { status: false, msg: "" },
    loading: false,
  });

  const classes = useStyles();

  useEffect(() => {
    setName(data.name);
    setDescription(data.description);
    setChannelID(data.id);
  }, [data]);

  const handleClose = () => {
    setOpen(false);
  };

  const formik = useFormik({
    initialValues: {
      description: !description ? "" : description,
      name: name,
      id: channelID,
    },
    enableReinitialize: true,
    validationSchema: channelValidationSchema,
    onSubmit: (values) => {
      updateChannelMetadata(values);
    },
  });

  const updateChannelMetadata = (values) => {
    setUpdateChannelAlert({
      ...updateChannelAlert,
      success: { status: false, msg: "" },
      error: { status: false, msg: "" },
      loading: true,
    });
    (async () => {
      try {
        const channelResponse = await editChannelMetadata(
          pubnubObject,
          values.name,
          values.description,
          channelID
        );
        handleClose();
        setUpdateChannelAlert({
          ...updateChannelAlert,
          success: { status: true, msg: "Channel updated successfully." },
          error: { status: false, msg: "" },
          loading: false,
        });
        channelUpdated(channelResponse);
      } catch (e) {
        handleClose();
        setUpdateChannelAlert({
          ...updateChannelAlert,
          success: { status: false, msg: "" },
          error: { status: true, msg: showError(e.status.errorData) },
          loading: false,
        });
      }
    })();
  };

  return (
    <div>
      <Dialog
        fullWidth={true}
        onClose={handleClose}
        aria-labelledby="customized-dialog-title"
        open={open}
        className={classes.modalLayout}
      >
        <DialogTitle id="customized-dialog-title" onClose={handleClose} disableTypography>
          <Typography testid="UpdateChannel" variant="h6" className={classes.headingFont}>
            Update Channel Metadata
          </Typography>
          {open ? (
            <IconButton
              testId="closeIcon"
              aria-label="close"
              onClick={handleClose}
              className={classes.closeButton}
            >
              <CloseIcon />
            </IconButton>
          ) : null}
        </DialogTitle>
        <AddChannelForm
          formik={formik}
          channelAlert={updateChannelAlert}
          handleClose={handleClose}
          buttonTitle={"UPDATE"}
          disabled={true}
        />
      </Dialog>
      {updateChannelAlert.error.status && (
        <SnackBar msg={updateChannelAlert.error.msg} status={"info"} />
      )}
      {updateChannelAlert.success.status && (
        <SnackBar msg={updateChannelAlert.success.msg} status={"success"} />
      )}
    </div>
  );
}
