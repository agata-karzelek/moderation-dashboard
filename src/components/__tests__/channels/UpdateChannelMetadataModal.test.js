import React from "react";
import { shallow } from "enzyme";
import UpdateChannelMetadataModal from "../../channels/UpdateChannelMetadataModal";

const mockSetOpenFn = jest.fn();
const mockData = {
  id: "one",
  name: "one",
  description: "",
  updated: "July 5th 2021, 4:18 pm",
  eTag: "AbWi2NmBocqbfg",
};

describe("Test Case For UpdateChannelMetadataModal", () => {
  let wrapper;
  beforeAll(() => {
    wrapper = shallow(
      <UpdateChannelMetadataModal
        open={true}
        data={mockData}
        setOpen={mockSetOpenFn}
        channelUpdated={jest.fn()}
      />
    );
  });

  test("Snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should render router", () => {
    expect(wrapper.find("AddChannelForm")).toHaveLength(1);
    expect(wrapper.find('[testid="UpdateChannel"]').getElements()[0].props.children).toBe(
      "Update Channel Metadata"
    );
  });

  test("check handleClose", () => {
    expect(mockSetOpenFn).toHaveBeenCalledTimes(0);
    wrapper.find('[testId="closeIcon"]').simulate("click");
    expect(mockSetOpenFn).toHaveBeenCalledTimes(1);
  });
});
