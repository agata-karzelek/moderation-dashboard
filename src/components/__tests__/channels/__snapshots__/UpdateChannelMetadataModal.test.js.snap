// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Case For UpdateChannelMetadataModal Snapshot 1`] = `
<div>
  <WithStyles(ForwardRef(Dialog))
    aria-labelledby="customized-dialog-title"
    className="makeStyles-modalLayout-8"
    fullWidth={true}
    onClose={[Function]}
    open={true}
  >
    <WithStyles(ForwardRef(DialogTitle))
      disableTypography={true}
      id="customized-dialog-title"
      onClose={[Function]}
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-headingFont-13"
        testid="UpdateChannel"
        variant="h6"
      >
        Update Channel Metadata
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(IconButton))
        aria-label="close"
        className="makeStyles-closeButton-3"
        onClick={[Function]}
        testId="closeIcon"
      >
        <Memo(ForwardRef(CloseIcon)) />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(DialogTitle))>
    <AddChannelForm
      buttonTitle="UPDATE"
      channelAlert={
        Object {
          "error": Object {
            "msg": "",
            "status": false,
          },
          "loading": false,
          "success": Object {
            "msg": "",
            "status": false,
          },
        }
      }
      disabled={true}
      formik={
        Object {
          "dirty": false,
          "errors": Object {},
          "getFieldHelpers": [Function],
          "getFieldMeta": [Function],
          "getFieldProps": [Function],
          "handleBlur": [Function],
          "handleChange": [Function],
          "handleReset": [Function],
          "handleSubmit": [Function],
          "initialErrors": Object {},
          "initialStatus": undefined,
          "initialTouched": Object {},
          "initialValues": Object {
            "description": "",
            "id": "",
            "name": "",
          },
          "isSubmitting": false,
          "isValid": true,
          "isValidating": false,
          "registerField": [Function],
          "resetForm": [Function],
          "setErrors": [Function],
          "setFieldError": [Function],
          "setFieldTouched": [Function],
          "setFieldValue": [Function],
          "setFormikState": [Function],
          "setStatus": [Function],
          "setSubmitting": [Function],
          "setTouched": [Function],
          "setValues": [Function],
          "status": undefined,
          "submitCount": 0,
          "submitForm": [Function],
          "touched": Object {},
          "unregisterField": [Function],
          "validateField": [Function],
          "validateForm": [Function],
          "validateOnBlur": true,
          "validateOnChange": true,
          "validateOnMount": false,
          "values": Object {
            "description": "",
            "id": "",
            "name": "",
          },
        }
      }
      handleClose={[Function]}
    />
  </WithStyles(ForwardRef(Dialog))>
</div>
`;
